### ISSUE TITLE
Feature request: option to create shell scripts with payload

### ISSUE THREAD

**Issue by sashamitrovich** [marklogic-community/ml-gradle#314]
Feature request: mlGradle should have an option to create shell scripts with payload instead of doing the actual deployment.

The background: many customers don't want to run gradle in production and tend to write shell scripts that use ML management-API. Having an option do automate creation of shell script based on mlGradle config would be very welcomed within such organisations as it would give them a level of control over what's exactly going on and ability to edit the scripts.

**Comment by rjrudin** [comment]
@sashamitrovich You mention two different problems:

1. Customers don't want to run Gradle in production
1. Customers would like the ability to edit the actual HTTP calls (with payloads) that are made to the Manage API

For the Gradle issue - are the customers have a JVM present in production, such that a non-Gradle solution that depends on a JVM would work? 

For the shell script issue - this is curious to me, because it implies that a package is built from a team's version control system, and then it's modified as part of a deployment to production. The Gradle properties plugin allows for overriding properties for an environment, but it sounds like here that the customer wants to edit payloads and possible even the set of HTTP calls itself. That doesn't sound like a good approach to me, because now what's being executed is different from what was built from version control. 

What kinds of edits are these customers looking to perform? I'd like to understand the use case better, particularly as generating shell scripts seems like one approach, and I can think of others that would externalize all of the calls so that they can be modified.

**Comment by grtjn** [comment]
I think the idea would be as follows: developers produce a release of whatever they built, which they hand over to sys-admins. That release would include a shell script + payloads to mldeploy the project on a target env, without needing java or gradle.

It is not as much that sys-admins would want to edit the payloads and scripts, but it gives a sense of control that they can with simple means, in case of whatever issue. It could be as simple as that they try to deploy against a different patch version, and need to edit one line in a config to make it work.

**Comment by hansenmc** [comment]
There is already a shell script to run, `gradlew`. And you can build with [ `--offline`](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:command_line_execution_options)

If they don't have a JVM installed, you could even adjust the build to include a local JVM in the packaged distro and reference that, instead of relying on a JVM being installed and available. 

Is the issue that they want to make the process for executing as easy as possible without any prerequisites, or is there an aversion to executing a JVM and Java/Groovy at all?

**Comment by rjrudin** [comment]
Hmm - there's a handful of places where "save" is called in ml-app-depoyer. Those could be consolidated to one method, and then something could be swapped in so that the "save" method writes to a file instead of making an HTTP call. 

There'd be some complications with authentication though - i.e. all the configuration of a Spring RestTemplate and an ML DatabaseClient would need to be mapped to curl command line arguments. Would also have to think through how e.g. passwords are handled. 

I think I could put together a very minimal cut of this for one resource - let's say for creating a privilege, which is about as simple as it gets. Can see how that works and iterate from there.

**Comment by sashamitrovich** [comment]
Thanks @grtjn for clarifying the process. Here's some background to why some customers don't want to run Gradle in production.

We're talking here about highly regulated industries - for each software they use there's an authority body that takes care there is an actual use case for that software, legally binding contracts, support SLAs, support contacts etc. The opposition to approving any new piece of software is enormous  - the IT landscape at some of our customers is already bloating with software (talking about 100s if not 1000s of different technologies that have to be understood an managed) plus our customers are mostly not software companies (software not being their product) so the culture of adopting best tools for the job, be it open source, free, commercial, whatever is not as developed as it is at MarkLogic or Google or Microsoft (it's the nature of business - they develop something else, software is not their focus).

We undergo the same scrutiny and sometimes this process of putting MarkLogic on the list of "approved and supported" software can take significant times within a big organisation. 

The same goes for any other kind of software, including Gradle - there needs to be a reason and willingness to approve it.

So, when we come in and tell them "oh by the way, we use Gradle to manage the MarkLogic configuration", the response (sometimes) is: "wait, that's a 3rd party software, right? We don't have permission to use it nor do we have trained people who know how to use it! Is there another way to manage MarkLogic without using unapproved 3rd party software or using 3rd party software that's already approved within our organisation?"

To summarise: it's not about whether they have JVM in production, it's even not about Gradle specifically - it's about personal liability of using not-approved software in production.

Thus this ticket - it's a solid workaround that gives the customer quality assurances (we, being the authors of mlGradle plugin, know how to create correct scripts to manage our software) and frees the customer up from personal liability for using unsanctioned software. Additionally - it gives them a piece of mind in case Gradle goes away one day (stops being maintained, supported, developed, whatever) - they have the scripts and they can build upon them.

Sorry for the long text. Does it answer your questions,  @rjrudin and @hansenmc?

**Comment by rjrudin** [comment]
In order to generate a shell script, the easiest thing I can think of is to run ml-gradle against an environment that matches the target environment. That's because ml-gradle makes a lot of GET calls to determine whether to create or update a resource. If one environment has the app deployed already but the other doesn't, then the scripts won't work because they'll be e.g. trying to update a resource instead of creating it. 

If that's possible, the underlying Spring RestTemplate can have a ClientHttpRequestInterceptor added to it to capture every call and write it out to a file that becomes the shell script.

Is that possible for your customer - running ml-gradle in a dev/test environment first to generate the script, and then running that script against the same/very similar environment in production?

Also, I assume curl is on the approved software list - what about a jar like mlcp, or if ml-gradle were jar'ed up so that it doesn't depend on Gradle?

**Comment by grtjn** [comment]
You need a shell script that works against any server. But you dont need to check existance. You invoke create, let it fail silently, and then do an update after that. If you are smart you invoke creation with minimal payload, and only send update with full payload..

**Comment by grtjn** [comment]
By the way, shell script allow for flow control too, if you want to do it more elegantly..

**Comment by grtjn** [comment]
I wanted to briefly touch up on my above remarks. I have worked for production departments at customers in the past, and we used to package up releases per target env specifically. Ideally you try to isolate most common settings like hosts, and ports, and maybe even make them easy to edit if really necessary, but sometimes you also have settings like a production flag that should be on or off. Or external services that MarkLogic should be able to reach. Settings you typically inject when deploying the code. Those can be baked hard into the release, provided you can recognize which package is for which target env..

It would obviously be convenient to make this as generic and flexible as possible, but in the end sys admins prefer pressing just one 'button' to make things happen..

**Comment by rjrudin** [comment]
See the attached script for a prototype. Just paste this at the bottom of your build.gradle file, and then run mlDeploy or any other task. It depends on ml-gradle 3.4.0 or higher - there's a comment in there for what to do if you're using 3.3.x or lower. 

[interceptor.txt](https://github.com/marklogic-community/ml-gradle/files/1782810/interceptor.txt)

**Comment by rjrudin** [comment]
Closing this for now, as my expectation is that folks interested in the above interceptor will try it out on their own project (since it can be customized easily), which will lead to a more generic solution that can be included by default in ml-app-deployer.
